src/core/gui/OpacityPreviewToolbox.cpp

```diff
@@ -0,0 +1,531 @@
+#include "OpacityPreviewToolbox.h"
+
+#include <cmath>  // for std::round
+
+#include <cairo.h>        // for cairo_set_operator, cairo_rectangle, cairo_...
+#include <glib-object.h>  // for G_CALLBACK, g_signal_connect
+#include <glib.h>         // for gdouble
+
+#include "control/Control.h"
+#include "control/ToolHandler.h"
+#include "gui/toolbarMenubar/ColorToolItem.h"
+#include "gui/toolbarMenubar/ToolMenuHandler.h"
+#include "util/Color.h"
+#include "util/raii/CairoWrappers.h"
+#include "util/raii/GObjectSPtr.h"
+
+#include "FloatingToolbox.h"
+#include "MainWindow.h"
+
+
+static int percentToByte(double percent) { return static_cast<int>(std::round(percent * 2.55)); }
+static double byteToPercent(int byte) { return byte / 2.55; }
+
+OpacityPreviewToolbox::OpacityPreviewToolbox(MainWindow* theMainWindow, GtkOverlay* overlay):
+        theMainWindow(theMainWindow), overlay(overlay, xoj::util::ref) {
+    this->opacityPreviewToolbox.widget = theMainWindow->get("opacityPreviewTool");
+
+    gtk_overlay_add_overlay(overlay, this->opacityPreviewToolbox.widget);
+    gtk_overlay_set_overlay_pass_through(overlay, this->opacityPreviewToolbox.widget, true);
+
+    g_signal_connect(overlay, "get-child-position", G_CALLBACK(this->getOverlayPosition), this);
+    g_signal_connect(theMainWindow->get("opacityPreviewToolScaleAlpha"), "change-value", G_CALLBACK(this->changeValue),
+                     this);
+
+    gtk_widget_add_events(this->opacityPreviewToolbox.widget, GDK_LEAVE_NOTIFY_MASK);
+    g_signal_connect(this->opacityPreviewToolbox.widget, "leave-notify-event", G_CALLBACK(this->leaveOpacityToolbox),
+                     this);
+}
+
+std::vector<OpacityPreviewToolbox::EventBox>::iterator OpacityPreviewToolbox::findEventBox(GtkWidget* eventBoxWidget) {
+    auto criteria = [eventBoxWidget](EventBox eventBox) { return eventBox.widget == eventBoxWidget; };
+    return std::find_if(this->eventBoxes.begin(), this->eventBoxes.end(), criteria);
+}
+
+gboolean OpacityPreviewToolbox::enterEventBox(GtkWidget* eventBoxWidget, GdkEventCrossing* event,
+                                              OpacityPreviewToolbox* self) {
+    auto result = self->findEventBox(eventBoxWidget);
+
+    if (result != self->eventBoxes.end()) {
+        self->showToolbox();
+        self->updateOpacityToolboxAllocation(*result);
+        self->opacityPreviewToolbox.eventBox = result;
+    }
+    return false;
+}
+
+bool OpacityPreviewToolbox::isPointerOverWidget(gint pointer_x_root, gint pointer_y_root, GtkWidget* widget) {
+    gint widget_x_root, widget_y_root;
+    if (GDK_IS_WINDOW(gtk_widget_get_window(widget))) {
+        gdk_window_get_origin(gtk_widget_get_window(widget), &widget_x_root, &widget_y_root);
+    }
```

Comment:
In older versions of my implementation, I was getting a warning about the widget not being a GDK_WINDOW, but I think that was related to missing calls of OpacityPreviewToolbox::update().

In any case, that test should be removed because that warning must reveal any problem in the implementation.
It should not be artificially hidden.